x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/EsqlFunctionRegistry.java

```diff
@@ -469,17 +472,24 @@ public List<String> argDescriptions() {
         }
     }
 
+    /**
+     * Build a list target data types, which is used by ImplicitCasting to convert string literals to a target data type.
+     */
     public static DataType getTargetType(String[] names) {
         List<DataType> types = new ArrayList<>();
         for (String name : names) {
-            types.add(DataType.fromEs(name));
-        }
-        if (types.contains(KEYWORD) || types.contains(TEXT)) {
-            return UNSUPPORTED;
+            DataType type = DataType.fromTypeName(name);
+            if (type != null && type != UNSUPPORTED) { // A type should not be null or UNSUPPORTED, just a sanity check here
+                // If the function takes strings as input, there is no need to cast a string literal to it.
+                // Return UNSUPPORTED, so that ImplicitCasting doesn't process it.
+                if (isString(type)) {
+                    return UNSUPPORTED;
```

Comment:
This is a strange way of saying that there shouldn't be an automatic conversion/casting for a specific data type.

Returning here a data type UNSUPPORTED can have double meaning:
- the actual target type is UNSUPPORTED
- the conversion shouldn't take place

I know this code existed before this PR, it might work and get the job done, but it's a confusing code; it is not clearly expressing the intention.

If this method - `getTargetType` - would have lived inside Analyzer and be `private` I would have been ok with it, but it's a public method, it can be accessed from anywhere in code and this means that this code duality I mentioned above can be confusing to anyone else wanting to re-use it.

Reply:
> This is a strange way of saying that there shouldn't be an automatic conversion/casting for a specific data type.
> 
> Returning here a data type UNSUPPORTED can have double meaning:
> 
> * the actual target type is UNSUPPORTED
> * the conversion shouldn't take place
> 

Understood, `UNSUPPORTED` is used to indicate that implicit casting is not supported for this argument here. I was thinking of using `NULL`, however it may lead people to interpret it as casting to a `NULL`. Before having a better solution, I'll add more comments here to indicate the purpose of using `UNSUPPORTED` here, perhaps we may want to add a flag in `ArgSignature` to indicate whether to do implicit casting for this argument or not, with a targetType if yes, but it adds one more attribute to `ArgSignature`, I'm not quite sure if it worths it or not.

> 
> If this method - `getTargetType` - would have lived inside Analyzer and be `private` I would have been ok with it, but it's a public method, it can be accessed from anywhere in code and this means that this code duality I mentioned above can be confusing to anyone else wanting to re-use it.

I can make `getTargetType` a private method in `EsqlFunctionRegistry`. Actually `Analyzer` doesn't call it, it is public because `AbstractFunctionTestCase` calls it. `AbstractFunctionTestCase` generates docs for automatically from annotations, however not all of the functions are registered in `EsqlFunctionRegistry`, like operators, so `AbstractFunctionTestCase` calls some `EsqlFunctionRegistry` methods separately to generate docs for them. I'll make it a private method here, as this target type is not externalized in the generated docs.